code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm

```diff
@@ -520,3 +514,158 @@
 	if(target.get_xeno_hivenumber() == owner.get_xeno_hivenumber())
 		return FALSE
 	return TRUE
+
+/datum/action/xeno_action/activable/punch/proc/add_to_combo(mob/living/A)
+	var/datum/action/xeno_action/activable/uppercut/uppercut_ability = owner.actions_by_path[/datum/action/xeno_action/activable/uppercut]
+	var/list/combo_list = uppercut_ability.combo
+	combo_list[A.name]++
+
+	if(combo_list[A.name] == WARRIOR_PERFECT_COMBO_THRESHOLD)
+		to_chat(A, span_highdanger("It looks like [owner] is predicting our movements, we should back off!"))
+
+	var/image/counter = image(null, A, null) //Visual clue for the warrior on how much combo each mob has
+	counter.maptext = MAPTEXT("[combo_list[A.name]]")
+	owner.client.images += counter
+
+	addtimer(CALLBACK(src,.proc/clear_counter, counter), WARRIOR_COMBO_FADEOUT_TIME)
+	addtimer(CALLBACK(uppercut_ability,/datum/action/xeno_action/activable/uppercut.proc/clear_combo, A, counter), WARRIOR_COMBO_FADEOUT_TIME, TIMER_OVERRIDE|TIMER_UNIQUE)
+
+/datum/action/xeno_action/activable/punch/proc/clear_counter(image/counter)
+	del(counter)
```

Comment:
del() usage is not allowed as it is extremely slow
instead please just remove all references to the image instead and it will get garbage collected